File: Algorithms/463. Island Perimeter/javascript/solution.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var islandPerimeter = function (grid) {
    let perimeter = 0;

    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[i].length; j++) {
            if (grid[i][j]) {
                perimeter += 4;

                if (i > 0 && grid[i - 1][j]) {
                    perimeter -= 2;
                }
                if (j > 0 && grid[i][j - 1]) {
                    perimeter -= 2;
                }
            }
        }
    }
    return perimeter;
};

File: Algorithms/463. Island Perimeter/swift/solution.swift
class Solution {
    func islandPerimeter(_ grid: [[Int]]) -> Int {
        var perimeter = 0
        for ridx in 0..<grid.count {
            for cidx in 0..<grid[0].count {
                if grid[ridx][cidx] == 0 { continue }

                let prev = cidx > 0 && grid[ridx][cidx - 1] == 1 ? 2 : 0
                let top = ridx > 0 && grid[ridx - 1][cidx ] == 1 ? 2 : 0

                perimeter += 4 - (prev + top)
            }
        }
        return perimeter
    }
}
